src/x/serialize/decoder_fast.go

```diff
@@ -27,40 +27,76 @@ import (
 
 // TagValueFromEncodedTagsFast returns a tag from a set of encoded tags without
 // any pooling required.
+
+type TagVisitor interface {
+	Length(n int)
+	Visit(tag, value []byte) bool
+}
+
 func TagValueFromEncodedTagsFast(
 	encodedTags []byte,
 	tagName []byte,
 ) ([]byte, bool, error) {
+	visitor := findTagValueVisitor{
+		tagName:  tagName,
+		tagValue: nil,
+	}
+	err := VisitTags(encodedTags, &visitor)
+	return visitor.tagValue, visitor.tagValue != nil, err
+}
+
+type findTagValueVisitor struct {
+	tagName  []byte
+	tagValue []byte
+}
+
+func (v *findTagValueVisitor) Length(_ int) {
+}
+
+func (v *findTagValueVisitor) Visit(tag, value []byte) bool {
+	if bytes.Equal(tag, v.tagName) {
+		v.tagValue = value
+		return false
+	}
+	return true
+}
+
+func VisitTags(
+	encodedTags []byte,
+	visitor TagVisitor,
```

Comment:
This is a nice pattern, but I'm afraid that using an interface here will cause allocation of any implementing struct (which we try to avoid at all cost in these code paths). You might be able to work around this by passing functions directly (instead of using an interface), but then the code will get uglier.
Let's see how it looks and performs (also I think you should check for the allocations, I think there are some benchmark/compiler options for this).